src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/core/ConnectedComponent.java

```diff
@@ -114,4 +117,10 @@ public void addLeafNode(Node node) {
     }
     return dependencyOrderedNodes;
   }
+
+  public Set<String> getNodeNames() {
+    getAllNodes();
```

Comment:
`getAllNodes()` is an expensive call. It should only be called if `nodeNames` is uninitialized.

Reply:
With current implementation, we are returning `nodeNames` without making the null check. If the `nodeNames` is not populated, then it should fail.